### ISSUE TITLE
Allow multiple Speculation Rules

### ISSUE THREAD

**Issue by midzer** [GoogleChromeLabs/quicklink#273]
**Is your feature request related to a problem? Please describe.**
According to https://developer.chrome.com/blog/prerender-pages/#multiple-speculation-rules the Speculation API allows multiple Speculation Rules. Right now `prerender: true` can only prerender a single URL in viewport.

**Describe the solution you'd like**
Remove https://github.com/GoogleChromeLabs/quicklink/blob/e6565ed12b4872b67a862b9fe2198fe0d73cb485/src/index.mjs#L247

**Additional context**
Have a nice weekend
midzer

**Comment by philwolstenholme** [comment]
I ran into this today too, based on the spec and https://developer.chrome.com/blog/prerender-pages I couldn't work out why we had the limit on one set of rules?

https://developer.chrome.com/blog/prerender-pages/#multiple-speculation-rules says:

> Multiple speculation rules can also be added to the same page, and they append to the existing rules

**Comment by philwolstenholme** [comment]
I noticed this today too, did earlier versions of the API only support one `script` element, perhaps?

**Comment by tunetheweb** [comment]
> did earlier versions of the API only support one script element, perhaps?

Yes that's correct. It's a relatively recent change (Chrome 107 I think) that allowed multiple prerenders.

**Comment by philwolstenholme** [comment]
I've had a look around the repo and made a few changes locally I think there are a few things we will need to do:

* Remove `isSpecRulesExists()` function definition and usage
* Remove `prerenderLimit` const definition
* Remove `if (toPrerender.size < prerenderLimit) {` check
* Add `toPrerender.clear();` to the 'reset' function returned from `listen()` (`toPrefetch.clear();` is there already)
* Call `toPrerender.clear();` after adding the Speculation Rules `script` elements to prevent a build up of the same URLs being included as new links are found on scroll (see example below)

Example of not clearing out the `toPrerender` set:

```html
<script type="speculationrules">
  { "prerender": [{ "source": "list", "urls": ["http://localhost:8080/work"] }] }
</script>
<script type="speculationrules">
  { "prerender": [{ "source": "list", "urls": ["http://localhost:8080/work", "http://localhost:8080/contact"] }] }
</script>
<script type="speculationrules">
  { "prerender": [{ "source": "list", "urls": ["http://localhost:8080/work", "http://localhost:8080/contact", "http://localhost:8080/github-stars"] }] }
</script>
<script type="speculationrules">
  { "prerender": [{ "source": "list", "urls": ["http://localhost:8080/work", "http://localhost:8080/contact", "http://localhost:8080/github-stars", "http://localhost:8080/github-stars?no-js"] }] }
</script>
```

**Comment by philwolstenholme** [comment]
All that said, I did quite quickly start seeing `Max number of prerendering exceeded`/`MaxNumOfRunningPrerendersExceeded` messages in the experimental `Prerender2` dev tools panel, so I'm thinking that Quicklink's approach of prerendering anything in the viewport might be a bit too resource intensive compared to only prerendering links that are being hovered over or `touchdown`'d on.

**Comment by tunetheweb** [comment]
Yeah there's current a limit of 10 prerenders. IMHO that is more than enough, and prerendering even that many (which I could easily see happening if prerendering all in-viewport links) seems like a lot to me.

**Comment by JiangWeixian** [comment]
`quicklink` export `prerender` API, currently, if call `prerender` multiple times, still throw this error. If user use `prerender` API manual, I think we should not limit prerender urls size.

Maybe should mv `isSpecRulesExists ` check function into `listen`.  🤔
